src/main/kotlin/org/ionproject/integration/file/implementation/AbstractFileDownloader.kt

```diff
@@ -0,0 +1,47 @@
+package org.ionproject.integration.file.implementation
+
+import java.io.File
+import java.net.HttpURLConnection
+import java.net.URL
+import java.nio.file.Path
+import org.ionproject.integration.file.`interface`.FileDownloader
+import org.ionproject.integration.file.exception.ServerErrorException
+import org.ionproject.integration.utils.Try
+
+abstract class AbstractFileDownloader() :
+    FileDownloader {
+
+    override fun download(urlStr: String, localDestination: String): Try<Path> {
+        if (urlStr.isEmpty() || localDestination.isEmpty()) {
+            return Try.ofError(IllegalArgumentException("Parameters url and localDestination need not be empty"))
+        }
+
+        val url = URL(urlStr)
+
+        val httpConn = Try.of(url.openConnection() as HttpURLConnection)
+
+        val responseCode = httpConn
+            .map { conn -> conn.responseCode }
+            .map { code -> throwIfServerError(code) }
+
+        val bytes = Try.map(responseCode, httpConn) { _, conn -> conn.inputStream.readBytes() }
+
+        val isValidFormat = bytes.map { byteArray -> checkFormat(byteArray) }
+
+        val file = isValidFormat.map { File(localDestination) }
+
+        Try.map(file, bytes) { f: File, b: ByteArray -> f.writeBytes(b) }
+
+        return file.map { f -> f.toPath() }
+    }
+
+    private fun throwIfServerError(code: Int): Int {
```

Comment:
We should throw, but instead return a `Try.Error`, right?

Reply:
I didn't understand... Should we throw or return Try.Error? Updating the method's return type to Try.Error isn't correct because whenever "code" is not in the referred range, the error code is returned. Also, as this method is called within a map, it should not return a Try, as I understand it, because the map would return Try<Try<>>, isn't it? Could you elucidate?